images/router/nginx/conf/nginx-config.template

```diff
@@ -5,91 +5,199 @@
+    {{ if or (eq $primaryCfg.TLSTermination "") (or (eq $primaryCfg.InsecureEdgeTerminationPolicy "Allow") (eq $primaryCfg.InsecureEdgeTerminationPolicy "Redirect")) }}
+    listen {{env "ROUTER_SERVICE_HTTP_PORT" "80"}};
+    {{ end }}
+  
+    server_name {{genCertificateHostName $cfgHost $primaryCfg.IsWildcard }};
+
+    {{ if eq $primaryCfg.InsecureEdgeTerminationPolicy "Redirect" }}
     if ($scheme = http) {
-        return   301 https://$host$request_uri;
+      return 301 https://$host$request_uri;
     }
-  {{ end -}}
- {{ end -}}
- 
-    access_log      /var/lib/nginx/logs/be_{{$cfgIdx}}.log main;
+    {{ end }}
 
-    location / {
+    {{- if (or (eq $primaryCfg.TLSTermination "edge") (eq $primaryCfg.TLSTermination "reencrypt")) -}}
+      {{ $cert := index $primaryCfg.Certificates $cfgHost -}}
+      {{ if ne $cert.Contents "" }}
+    listen 127.0.0.1:{{env "ROUTER_SERVICE_SNI_PORT" "10444"}} ssl;
+    ssl_certificate     {{$workingDir}}/certs/{{$primaryCfgIdx}}.pem;
+    ssl_certificate_key {{$workingDir}}/certs/{{$primaryCfgIdx}}.pem;
+     {{ end }}
+    {{ end }}
+
+    access_log      /var/lib/nginx/logs/route_{{$cfgHost}}.log main;
+
+    {{ range $cfgIdx, $cfg := $configs }}
+    # this path belongs to the {{ $cfgIdx }} route
+    location {{ if eq $cfg.Path ""}}/{{ else }}{{$cfg.Path}}{{ end }} {
+      {{ if eq $cfg.TLSTermination "reencrypt" }}
+      proxy_ssl_name {{ $cfg.Host }};
+        {{- if gt (len (index $cfg.Certificates (printf "%s_pod" $cfg.Host)).Contents) 0 }}
+      proxy_ssl_trusted_certificate {{ $workingDir }}/cacerts/{{$cfgIdx}}.pem;
+      proxy_ssl_verify on;
+        {{ else }}
+          {{ if gt (len $defaultDestinationCA) 0 }}
+      proxy_ssl_trusted_certificate {{ $defaultDestinationCA }};
+      proxy_ssl_verify on;
+          {{ else }}
+      proxy_ssl_verify off;
+          {{ end }}
+       {{ end }}
+      proxy_pass      https://be_{{$cfg.Namespace}}_{{$cfg.Name}};
+      {{ else }}
       proxy_pass      http://be_{{$cfg.Namespace}}_{{$cfg.Name}};
+      {{ end }}
     }
+    {{ end }}
+  }
+{{ end -}}
+}
+
+stream {
+  log_format basic '$remote_addr [$time_local] '
+                     '$protocol $status $bytes_sent $bytes_received '
+                     '$session_time "$ssl_preread_server_name" $dest';
+  
+  upstream https-route {
+    server 127.0.0.1:{{env "ROUTER_SERVICE_SNI_PORT" "10444"}};
+  }
+
+  {{- range $cfgIdx, $cfg := $passthroughAliases }}
+  upstream be_passthrough_{{$cfg.Namespace}}_{{$cfg.Name}} {
+      {{ if ne $cfg.ActiveEndpoints 0 }}
+        {{- range $serviceUnitName, $weight := $cfg.ServiceUnitNames }}
+          {{- with $serviceUnit := index $.ServiceUnits $serviceUnitName }}
+            {{- range $idx, $endpoint := endpointsForAlias $cfg $serviceUnit }}
+    server {{$endpoint.IP}}:{{$endpoint.Port}};
+            {{ end -}}
+          {{ end -}}
+        {{ end -}}
+      {{ else }}
+    server 127.0.0.1:{{env "ROUTER_SERVICE_UNREACHABLE_PORT" "10088"}};
```

Comment:
since it is a TCP upstream, we cannot return the 503 code. again, the dummy server is a hack to overcome the NGINX requirement of having at least 1 server in an upstream.